### ISSUE TITLE
Error while running ./script/prepdocs.sh

### ISSUE THREAD

**Issue by AxelQSam** [Azure-Samples/azure-search-openai-demo#1415]
### This issue is for a: (mark with an `x`)
```
- [X] bug report -> please search issues before submitting
- [ ] feature request
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)
```

### Minimal steps to reproduce
>Remove blob with --removeall in prepdocs.sh. Remove files from folder "data". Add new data to the file and then rerun prepdocs.sh. 

### Any log messages given by the failure

Traceback (most recent call last):
  File "/workspaces/SAM-Azure-AI-Search/./scripts/prepdocs.py", line 426, in <module>
    loop.run_until_complete(main(ingestion_strategy, azd_credential, args))
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 654, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/workspaces/SAM-Azure-AI-Search/./scripts/prepdocs.py", line 244, in main
    await strategy.run(search_info)
  File "/workspaces/SAM-Azure-AI-Search/scripts/prepdocslib/filestrategy.py", line 76, in run
    await search_manager.update_content(sections, blob_image_embeddings)
  File "/workspaces/SAM-Azure-AI-Search/scripts/prepdocslib/searchmanager.py", line 215, in update_content
    embeddings = await self.embeddings.create_embeddings(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/SAM-Azure-AI-Search/scripts/prepdocslib/embeddings.py", line 119, in create_embeddings
    return await self.create_embedding_batch(texts)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/SAM-Azure-AI-Search/scripts/prepdocslib/embeddings.py", line 90, in create_embedding_batch
    async for attempt in AsyncRetrying(
  File "/workspaces/SAM-Azure-AI-Search/scripts/.venv/lib/python3.11/site-packages/tenacity/_asyncio.py", line 71, in __anext__
    do = self.iter(retry_state=self._retry_state)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/SAM-Azure-AI-Search/scripts/.venv/lib/python3.11/site-packages/tenacity/__init__.py", line 314, in iter
    return fut.result()
           ^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/workspaces/SAM-Azure-AI-Search/scripts/prepdocslib/embeddings.py", line 97, in create_embedding_batch
    emb_response = await client.embeddings.create(model=self.open_ai_model_name, input=batch.texts)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/SAM-Azure-AI-Search/scripts/.venv/lib/python3.11/site-packages/openai/resources/embeddings.py", line 214, in create
    return await self._post(
           ^^^^^^^^^^^^^^^^^
  File "/workspaces/SAM-Azure-AI-Search/scripts/.venv/lib/python3.11/site-packages/openai/_base_client.py", line 1725, in post
    return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/SAM-Azure-AI-Search/scripts/.venv/lib/python3.11/site-packages/openai/_base_client.py", line 1428, in request
    return await self._request(
           ^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/SAM-Azure-AI-Search/scripts/.venv/lib/python3.11/site-packages/openai/_base_client.py", line 1519, in _request
    raise self._make_status_error_from_response(err.response) from None
openai.BadRequestError: Error code: 400 - {'error': {'message': "'$.input' is invalid. Please check the API reference: https://platform.openai.com/docs/api-reference.", 'type': 'invalid_request_error', 'param': None, 'code': None}}

### Expected/desired behavior
>

### OS and Version?
> Locally run with dev container (linux).

### azd version?
azd version 1.7.0

### Versions
>

### Mention any other details that might be useful
Will start and extracting data in batches a few times before error. File type is pdf and size 1.6mb.

Searching index with * only qives:
```
{
  "@odata.context": "https://gptkb-you7tt32egq5k.search.windows.net/indexes('gptkbindex')/$metadata#docs(*)",
  "value": []
}
```
<img width="1053" alt="Skärmbild 2024-03-13 115358" src="https://github.com/Azure-Samples/azure-search-openai-demo/assets/162345082/2d676c8c-98d3-4339-88d4-025481facda0">

**Comment by OrionSehn** [comment]
I can confirm I also have a very similar result, I think on a commit a little back:
![image](https://github.com/Azure-Samples/azure-search-openai-demo/assets/97123736/684381b2-767a-442b-96c7-bac7c7edcff7)

Commit ID: 
commit 68e78df005fcfaec0859e85d9f341efe7e2e6418 (HEAD -> main, tag: 2024-03-07)

This seems to occur when you run either with new data, or from fresh.

**Comment by pamelafox** [comment]
Hm, that commit only introduced a new parser for text files, so it shouldn't affect PDF files. I wonder if it had to do with the splitter changes that were fairly recent. Is it always repeatable with the same PDF? If any of you have a non-sensitive PDF that you can email me to replicate the issue, that'd be great. pamelafox @ microsoft

**Comment by AxelQSam** [comment]
> Hm, that commit only introduced a new parser for text files, so it shouldn't affect PDF files. I wonder if it had to do with the splitter changes that were fairly recent. Is it always repeatable with the same PDF? If any of you have a non-sensitive PDF that you can email me to replicate the issue, that'd be great. pamelafox @ microsoft

Emailed you the following and a pdf:
I found a new pdf that is non-sensitive that I could replicate the issue. Tried with multiple PDFs and small pdfs doesn't seem to have a problem (around 1mb and 4 slides) but as soon as it needs to batch more than 2-5 times it crashes.

**Comment by n-y-kim** [comment]
Having a same issue here. Would rolling back to another commit would resolve this issue temporarily?

**Comment by OrionSehn** [comment]
I think it ends up generating a .pdf.md5 for the ones which it does not work for, and then does not actually index those problematic pdfs, as it skips them not detecting any changes. So when you repeat this multiple times you don't actually perform the indexing of those documents, and end up skipping them.

**Comment by pamelafox** [comment]
Yeah unfortunately we write the md5 file optimistically. There's a PR to use a hash on the blob storage instead, but that PR isn't ready.

I have replicated the issue. As a workaround, you can disable batch vectors by passing `--disablebatchvectors` to the prepdocs.py arguments in prepdocs.sh/prepdocs.ps1. First delete the md5 file manually for any file that failed.

**Comment by pamelafox** [comment]
(I am working to understand the root cause, so hopefully will have a fix today or tomorrow).

**Comment by pamelafox** [comment]
Update: This error is happening when we pass a text of length 0 (an empty string) to the batch embeddings API. The single embedding API is fine with that input, but the batch embedding API is not. (See https://github.com/openai/openai-python/issues/576)

Now, I don't know yet why we have sections that have 0 text in them, as I don't expect that to happen in most cases (possibly for GPT4-vision, but this occurs with vision disabled as well). I'm going to ask @tonybaloney to see if it was related to the recent splitting change.

As another workaround, you can put this code in create_embedding_batch:

```
# replace any empty strings with whitespace for now
batch.texts = [text if text else " " for text in batch.texts]
emb_response = await client.embeddings.create(model=self.open_ai_model_name, input=batch.texts)
```

The batch embedding endpoint seems fine with a whitespace string.

**Comment by pamelafox** [comment]
This is now fixed! Please pull main or changes from #1423
